### DISCUSSION TITLE
Extending the visualiser to make it a proper workbench

### DISCUSSION THREAD

**Discussion by sgenoud** [sgenoud/replicad#69] [Visualiser]
You can now use an editor online with the visualiser to make it into a "workbench".

<img width="1402" alt="Capture d’écran 2023-02-10 à 14 38 00" src="https://user-images.githubusercontent.com/263325/218105898-df997a8c-8f1f-44bd-80cc-ab50b33ac70f.png">

So what does it bring with it? One thing is now it is easier to share a simple model [with just a link](http://studio.replicad.xyz/workbench?code=UEsDBAoAAAAIANRsSlZhI%252Bx9gAIAAOsEAAAHAAAAY29kZS5qc21UUW%252FaMBB%252B51eceHJY6gAFKlqBuqnr27Sq28Mq1E4muUBUY0eO04I6%252FvvOdpJSqVKU2L7vvvN9%252FpxkABup10KCwVIWqchgkPSSwQCu7aHEDHN4AzfSORS7UhvL%252Bi2yH8HxQ1ovICjjBqWolTgQIDd6B%252F2ttWV1mSRppnj1fChF%252BswzfEmy87OswV7P%252Ble9XqpV5RgyI15juKF3oTaxn97rWmWY3WNqhdpIJPJFV59S81qlttDKg%252B%252B0PGy0uqXqd7pQtmKl%252F0Tw1gMIVdZlRQwO3QRXw8foisJhxitiRjaKeK7Nd5FuGSsjWCxdHpeFwt%252BaFnyCQVsb5QOp1BUyWj22zRjaLVIhpnz2V2PEgTtZ2BtIVBu7vQQFxxjY3xgKjykov4d7r2ZHkundL8TMMVX09UDXjEQLglbd4slmOjlY6BngRRiwLl3AlwUM97NsvL6Y51PfAvjQD2G3vNjVkll4Anotl0sYTaOYov9g1CGfFvT68jn8okHPWnizIfaBcxJFfjCMIIHJeD6Zzy7G85nLOJJ47%252F13fexEodjp%252BQVNnAk6cdhofD7xZQMknGSAbJCNhxHfiZIxL94q5NFkAKPhMIaT%252BdRb4YTH28kxtd4OR9i4yvfJX7TRShdsRVz0eE5HFKIpStnYsmKB3B2dW3a8Cl9bvzPfAbkOmEM0xYHuYLuP9kjD7nRt0dyKFBs7f2Z%252Bv9IcSFvSf3lek2E7Cq7zvELLzsZROxzxabD58V2QVYqKMv7EEAYPjx3f2t1T6uKb3vMQ9L02LvCYIIglvSspLLKzjq0ZPTSaFW5S0Y1nn%252F0BWKOwkzqK6MeCNt3%252BVHdSKLqCHPfW1Jm7we42%252FgdQSwECFAAKAAAACADUbEpWYSPsfYACAADrBAAABwAAAAAAAAAAAAAAAAAAAAAAY29kZS5qc1BLBQYAAAAAAQABADUAAAClAgAAAAA%253D). (note that the studio needs to update to allow for the workbench to be functional).

You can also use a "reload a file from you disk mode" by clicking on the reload button in the editor pane.

Shared models now also have an edit link to automatically open within the workbench.

I am curious if you find this useful!

**Comment by BarbourSmith** [comment]
I love it, it's fantastic. Especially when looking at the examples in the docs it's great to be able to click the "edit" button and try them out directly.

**Reply by sgenoud** [reply]
Thanks, I have made some additional work on the documentation as well. You can now click on a "workbench" button to open in the workbench and start editing.

I used it extensively for the [new tutorial example](https://replicad.xyz/docs/tutorial-making-a-watering-can/intro)

**Comment by tve** [comment]
This is essential for me. I haven't used replicad any other way. I'd love to see the whole thing enhanced!

(Am willing to contribute, if I can. Is the source code for the visualizer & workbench open?)

**Reply by sgenoud** [reply]
Thanks!

I definitely should open source the whole thing, I "just" need to find the time now 😉 

P.S. I will answer your other comments and question - it has been a busy weekend and I'll look into them as soon as I have the bandwidth!

**Reply by tve** [reply]
Great to hear! I was pondering that without source for the visualizer I can't troubleshoot or fix anything and that would end up making it a no-go.

**Comment by tve** [comment]
Relatively simple wish list:
- insert a delay (400ms?) before re-evaluating the model and keep delaying if more keys are pressed: I have a complex model and it takes several seconds to update, so when I make a change it gets stuck for seconds on the first char I type which always results in a syntax error or a kernel error
- when there is an error, don't show the amoeba animation, instead show the model, perhaps with a red-grey transparent overlay or something like that so I can still see what I had, I do rely on the visual while I type changes and having the visual go away as soon as I type the first char is really annoying
- don't remove the XY plane grid when zooming in close, I do use it to check dimensions
- don't reset the view when updating, instead provide a button to reset it

More complex wish list:
- provide a way to measure things in the visualizer

**Reply by raydeleu** [reply]
Last week I worked with the workbench and I can confirm many items in your wish list. I would even prefer the removal of the automatic update and only re-evaluate the code on request. Furthermore I noticed that the amoeba animation seems to take a lot of computing power. The fan of my Macbook Air was often blowing at full speed during this animation which did not help to reduce my frustration after having made a mistake. 

I noticed that when you select an element in the workbench it lists the type, id and position of this element. Perhaps this can help instead of measuring? However, I also noticed that the dimensions are not always accurate enough. For example, if you select an edge and use this dimension in a ```containsPoint``` finder for edges, this often fails when the dimension is the result of an earlier cut operation. It only seems to work with dimensions that are specified exactly by the user. I tried to clarify this in the image below.  

![example_dimensions](https://github.com/sgenoud/replicad/assets/38007983/e0131a1a-dede-47ad-aec8-ec42716c824f)

**Reply by sgenoud** [reply]
Yes, the value shown is rounded to be more readable. It might be nice on my part to add a midpoint as well. You can also use the `atDistance` method to look for a non exact point (and have a distance of 1e-2 for instance).

**Reply by raydeleu** [reply]
I tried the selection with ```atDistance``` but this does not work in all situations. When I start with a newly defined cube it indeed seems to work for an error of up to 0.01 mm. But in the example below I tried to use the coordinates as displayed, 

![image](https://github.com/sgenoud/replicad/assets/38007983/609ad238-7ba3-44f3-a896-86ada8815bbb)

``` javascript
 let shapeRounded = shapeUnrounded
  //.fillet(1.0,(e)=>e.atDistance(0 , [0.00, 23.93, 15.90]))
  //.fillet(1.0,(e)=>e.containsPoint([0, ly*ycut/2 , lz+2*th]))
````

The version with ```containsPoint``` works as the coordinates are known exactly. The version with ```atDistance``` does not work. The coordinates I used, [0.00, 23.93, 15.90], should represent the middle of the highlighted edge. Wouldn't it be possible to use a selection like "the element that is closest to the referenced point"? I have the impression that this is how https://mirmik.github.io/zencad/en/fillet.html solved it. 

Of course the downside to using hardcoded coordinates is that the part is no longer parametric. Any change to the design makes this approach fail. The same can be true for using edge indices although these should remain pretty constant as long as the design of the shape does not change (and with it the number of edges and/or faces).
